optimum/exporters/onnx/model_patcher.py

```diff
@@ -448,3 +451,42 @@ def __init__(
             self._patch_func = _prepare_decoder_attention_mask
             self._orig_func_name = "_prepare_decoder_attention_mask"
             self._orig_func = self._model_to_patch._prepare_decoder_attention_mask
+
+
+if is_open_clip_available():
+    import open_clip
+
+def _text_global_pool_patched(x, text: Optional[torch.Tensor] = None, pool_type: str = 'argmax'):
+    if pool_type == 'first':
+        pooled, tokens = x[:, 0], x[:, 1:]
+    elif pool_type == 'last':
+        pooled, tokens = x[:, -1], x[:, :-1]
+    elif pool_type == 'argmax':
+        text = text.to(dtype=torch.int32)  # ONNX Runtime is unable to run argmax with int64 input, hence this cast.
+        # take features from the eot embedding (eot_token is the highest number in each sequence)
+        assert text is not None
+        pooled, tokens = x[torch.arange(x.shape[0]), text.argmax(dim=-1)], x
+    else:
+        pooled = tokens = x
+
+    return pooled, tokens
+
+
+class OpenCLIPModelPatcher(CausalAttentionMaskModelPatcher):
+    def __init__(
+        self,
+        config: "OnnxConfig",
+        model: Union["PreTrainedModel", "TFPreTrainedModel"],
+        model_kwargs: Optional[Dict[str, Any]] = None,
+    ):
+        super().__init__(config, model, model_kwargs)
+        self.original_text_global_pool = open_clip.transformer.text_global_pool
+
+    def __enter__(self):
+        self.patch_ops()
+        open_clip.transformer.text_global_pool = _text_global_pool_patched
```

Comment:
Can you try:
```suggestion
        open_clip.transformer.text_global_pool.__code__ = _text_global_pool_patched.__code__
```

I believe that given that the model is already instanciated, the old `text_global_pool` is still used despite patching. This can be checked by printing stuff in `_text_global_pool_patched`.

Reply:
Thanks! Yep with the addition of `__code__`, print stmt works within `_text_global_pool_patched`. The `torch.export` seems to have run without issues, and the onnx input names are now `['images','text']`. Later this error is raised:
```
  File "/Users/isaacchung/work/optimum/optimum/exporters/onnx/base.py", line 331, in fix_dynamic_axes
    outputs = session.run(None, onnx_inputs)
  File "/Users/isaacchung/virtualenv/v1/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 216, in run
    self._validate_input(list(input_feed.keys()))
  File "/Users/isaacchung/virtualenv/v1/lib/python3.9/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 198, in _validate_input
    raise ValueError(
ValueError: Required inputs (['image', 'text']) are missing from input feed (['input_ids', 'pixel_values', 'attention_mask']).
```
Interestingly, in `fix_dynamic_axes`, setting `to_fix = []` allows the ONNX export to succeed, but the following warning is raised:
```
The ONNX export succeeded with the warning: The exported ONNX model does not have the exact same outputs as what is provided in OpenCLIPOnnxConfig. Difference: 4624.
```

Reply:
Yes, the input name should be fixed as well there. You can check how the inputs are generated and whether the input remap is called & where/why/why not.

Reply:
Made some good progress in this new commit. I feel that we are very close. The only diff now is in the values are mismatching:
```
Validating ONNX model clip_onnx/model.onnx...
        -[✓] ONNX model output names match reference model (logit_scale, image_features, text_features)
        - Validating ONNX Model output "text_features":
                -[✓] (2, 512) matches (2, 512)
                -[x] values not close enough, max diff: 0.3609406352043152 (atol: 1e-05)
        - Validating ONNX Model output "image_features":
                -[✓] (2, 512) matches (2, 512)
                -[x] values not close enough, max diff: 0.36094051599502563 (atol: 1e-05)
        - Validating ONNX Model output "logit_scale":
                -[✓] () matches ()
                -[✓] all values close (atol: 1e-05)
The ONNX export succeeded with the warning: The maximum absolute difference between the output of the reference model and the ONNX exported model is not within the set tolerance 1e-05:
- text_features: max diff = 0.3609406352043152
- image_features: max diff = 0.36094051599502563.
 The exported model was saved at: clip_onnx
```

Reply:
Awesome, great work! I wonder if 0.36 is actually small compared to the output distribution or not. It could be worth checking the mean absolute difference. Maybe there's a dropout with `training=True` somewhere (like here https://github.com/huggingface/transformers/blob/0baa9246cb1ddac355db1df7824a521426599eb7/src/transformers/models/speecht5/modeling_speecht5.py#L720, though I doubt it)

Reply:
Thanks! The mean absolute difference is around 0.043, so maybe 0.36 is quite significant.
```
(Pdb) np.mean(np.abs(ref_value - ort_value))
0.042761594
```
The max diff value changes slightly at every run, but I think that's likely due to the random dummy inputs. Dropout is set to be 0 in the configs too. I initially ran these on a CPU, so trying now on a GPU.

Reply:
Same error - so it does not seem to be a device issue.

Reply:
I'm actually working on exporting open_clip models and ran into an issue when calling `torch.jit.trace`. Not sure if it's the same issue, but I had to explicitly do

```
for param in model.parameters():
   param._requires_grad(False)
```

`.eval()` and `with torch.no_grad()` didn't help; it would get stuck while exporting unless I added the above snippet.

Reply:
Where did you add the snippet btw?

Reply:
Just before calling `torch.jit.trace`. [Here's the link](https://github.com/immich-app/immich/commit/b5badc07acfbdc934cef13d2689f16af1acf0f38#diff-3951c3cdd96468dc58826b8555b35b0216a5e61d2fa224298a1b97fe77999326R57)

Reply:
Thanks. Just tested it out by add that within `TasksManager.get_model_from_task`, and setting `jit=False`, still the same error. I don't think the issue here is jit related.